Don't create a statusbar in Qt, wx backends.

anntzer opened the pull request:
Display the coordinates text on the right of the toolbar, consistently
with the GTK3, wx, and (I think?) macosx backends.  This helps when
embedding Matplotlib in larger GUIs, because Matplotlib may not
necessarily control the statusbar (which is typically global for the
whole window) but controls the toolbar if there's one.

If we decide to go this route we could probably also kill the status bar
for toolmanager.

xref https://github.com/matplotlib/matplotlib/issues/17085.

## PR Summary

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

tacaswell commented on doc/api/api_changes_3.3/behaviour.rst:
```diff
@@ -163,3 +163,9 @@ The default method used to format `.Slider` values has been changed to use a
+
+Qt and wx backends no longer create a status bar by default
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The coordinates information is not displayed in the toolbar, consistently with
```
```suggestion
The coordinates information is now displayed in the toolbar, consistently with
```
?

anntzer replied:
Yup, just a typo.

tacaswell replied:
but a very confusing typo!

tacaswell commented on doc/api/api_changes_3.3/behaviour.rst:
```diff
@@ -163,3 +163,9 @@ The default method used to format `.Slider` values has been changed to use a
+
+Qt and wx backends no longer create a status bar by default
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The coordinates information is now displayed in the toolbar, consistently with
```
It looks like for Qt we never created the status bar, we just injected an extra widget into it?

anntzer replied:
We did -- just calling statusBar() was enough to create one implicitly.

timhoffm commented on lib/matplotlib/backends/backend_qt5.py:
```diff
@@ -555,23 +555,21 @@ def __init__(self, canvas, num):
                 self.statusbar = StatusbarQt(self.window, self.toolmanager)
+                sb_height = self.statusbar.sizeHint().height()
+        else:
+            sbs_height = 0
```
Typo? Is this the same as `sb_height`?

anntzer replied:
actually they should all be sbs_height, fixed

anntzer commented:
> The coordinates were shown in the status bar in the bottom left. Now they move to the top right in the toolbar, which may be surprising for users.

In general I know this may be a bit too disruptive to be worth it -- that's why I tagged as "needs comments/discussion".  But I'm not sure there's much for me to argue about here, other than "I don't think it's that bad?"

> semi-related: Some backends have the toolbar at the top of the figure window, some at the bottom. This is also confusing.

Well, fixing that would be even more disruptive :-)

> The coordinates clunch to the top right corner of the status bar with an arbitrary margin to the right. This looks really ugly (personal taste, but really).

Looks like it was just a matter of `.rstrip()`ing the string (done), does that help for you?
￼
> If one make the figure a little smaller, the coordinates are not displayed anymore because the label is clipped

- This has always been the case for gtk3/tk, which are admittedly less used than qt (although tk was/is(?) the default on C. Gohlke's Windows wheels) but still this suggests this isn't too much of a real problem.
- #16776 (wink) will make things better by shortening the cursor texts, removing non-significant digits.
- I pushed a separate commit with two more things that help here: 1) move the image pixel value to a second line (this particularly helps if the image uses RGB values (i.e. 3 floats); 2) don't display the "mode" (zoom/pan), which is now quite redundant with the toolbar button which should appear as being depressed on the same toolbar.

timhoffm commented:
> > [...] which may be surprising for users.

> In general I know this may be a bit too disruptive to be worth it -- that's why I tagged as "needs comments/discussion".  But I'm not sure there's much for me to argue about here, other than "I don't think it's that bad?"

Probably. I'm not categorically opposed, but we should consider this in our decision.

> > The coordinates [...] looks really ugly (personal taste, but really).

> Looks like it was just a matter of `.rstrip()`ing the string (done), does that help for you?

Yes, that's better, but only half of it. Can we center the label vertically? I don't see a reason  why we need to align it to the top (`backend_qt5.py l.716`). For multi-line text the label will expand in both directions, so two lines should still be fine. There's only an issue for more lines, which then would show only the middle section, not the beginning i guess (not tested). If one wanted to be 100% save, one could change the alignment to center for single lines and top for multi line texts.

> - #16776 (wink) will make things better by shortening the cursor texts, removing non-significant digits.

Merged. Please rebase so that I can test that more easily :smile:.

> - I pushed a separate commit with two more things that help here: 1) move the image pixel value to a second line (this particularly helps if the image uses RGB values (i.e. 3 floats); 2) don't display the "mode" (zoom/pan), which is now quite redundant with the toolbar button which should appear as being depressed on the same toolbar.

looks good.

anntzer commented:
Rebased.

> Yes, that's better, but only half of it. Can we center the label vertically? I don't see a reason why we need to align it to the top (backend_qt5.py l.716). For multi-line text the label will expand in both directions, so two lines should still be fine. There's only an issue for more lines, which then would show only the middle section, not the beginning i guess (not tested). If one wanted to be 100% save, one could change the alignment to center for single lines and top for multi line texts.

Agreed centered looks better; changed that.
I don't think we ever manually set the message to more than two lines (or even more than one line before this PR -- it used to go to the status bar after all, so probably setting more than one line wasn't supported at all anyways); a quick check shows that setting the text to more than two lines just shows the first two lines and crops the rest.

anntzer commented:
Also, I think in the wx case, this is "clearly" an esthetic improvement (sure, it's all subjective, but heh):

before
![old](https://user-images.githubusercontent.com/1322974/79042921-fa071f80-7bfb-11ea-844b-876cd4f03079.png)

after
![new](https://user-images.githubusercontent.com/1322974/79042919-f83d5c00-7bfb-11ea-8e5b-da5414229f84.png)

timhoffm commented on lib/matplotlib/backends/backend_qt5.py:
```diff
@@ -714,8 +712,7 @@ def _init_toolbar(self):
             self.locLabel = QtWidgets.QLabel("", self)
-            self.locLabel.setAlignment(
-                    QtCore.Qt.AlignRight | QtCore.Qt.AlignTop)
+            self.locLabel.setAlignment(QtCore.Qt.AlignRight)
```
This doesn't change the alignment for me. I have to explicitly set
```suggestion
            self.locLabel.setAlignment(QtCore.Qt.AlignRight | QtCore.Qt.AlignCenter)
```

anntzer replied:
I think this should be AlignVCenter, sure, that's fine.

timhoffm commented:
Out of curiosity: What are your system settings so that the wx toolbar is dark. I'm on Kubuntu with a dark theme. The Qt toolbar is dark as expected, but the wx toolbar is still light and also has larger icon margins:

![image](https://user-images.githubusercontent.com/2836374/79045872-ce8d3080-7c0d-11ea-8e9b-debf3ab9cc24.png)

anntzer commented:
Arch + Breeze Dark (also for GTK, per https://wiki.archlinux.org/index.php/Uniform_look_for_Qt_and_GTK_applications#Styles_for_both_Qt_and_GTK).

Probably would be nice to have the same thing in wx to make buttons light when on a dark background as you did for Qt (that was great :-)), but I don't use wx often enough for this to be worth the trouble...

timhoffm commented:
> Probably would be nice to have the same thing in wx to make buttons light when on a dark background as you did for Qt (that was great :-))

I was about to write that, but couldn't get the dark backfround to test.

ImportanceOfBeingErnest commented:
One reason to not use the tk backend for me was always that it resizes the figure due to the coordinates being shown inside the navigation bar.

![tkresize](https://user-images.githubusercontent.com/23121882/79047385-86264080-7c16-11ea-91f2-a0847e5f32a4.gif)

Now even if no resizing occurs in Qt, the fact that the coordinates are not visible for small figures is not desireable.

anntzer commented:
@timhoffm If you want to fix icon colors for dark themes + gtk, I have a mostly working patch at https://gist.github.com/anntzer/8909837f8438210538336651a87aed9c; feel free to adopt it.

@ImportanceOfBeingErnest I'm a bit confused by what you're saying about resizes due to the coordinates being shown?  Most often I'd assume you use the mouse to resize the window, in which case the cursor is not over the canvas and thus no coordinates are being shown?  I agree the situation with very small windows is perhaps not great.

OTOH the other solutions are either 1) to also move the text to the statusbar for tk/gtk/macosx (not sure for the last one), together with, hopefully, a consistent decision as to what to do in the embedding case where we may or may not control the statusbar; or 2) to just not bother about consistency here at all.

ImportanceOfBeingErnest commented:
>  I'm a bit confused by what you're saying about resizes due to the coordinates being shown? Most often I'd assume you use the mouse to resize the window, in which case the cursor is not over the canvas and thus no coordinates are being shown?

Maybe it's not actually obvious from the screen capture above, but the problem is that the window and hence the figure *resizes itself* because while the mouse is over the figure, the coordinates change and thereby the width of the nav bar.

anntzer commented:
Ah, but that just looks like a bug?  The window should just keep its size and crop the text as appropriate?

ImportanceOfBeingErnest commented:
Might be a bug, but if the text is cropped, you have no chance of reading the text.

anntzer commented:
Fair enough, but I cannot think of another program which changes its window size just because a label would be cropped.

Thinking about it again, another option would be to switch to a tooltip when the label would be cropped...

timhoffm commented:
Is it really that bad if the coordinates are not displayed for small plots? I'd argue that you don't have much precision in small plots anyway so the benefit would be limited. But tooltip might be good.

ImportanceOfBeingErnest commented:
Yes, I consider it as bad, if just because you have a small figure, you cannot see the coordinates any more.

A tooltip may be a solution, just maybe not *within* the axes, because that would hide the data beneath it. Maybe at the lower corner of the figure, or even within the navigation bar (which will not be used anyways while having the mouse in the figure). 
![image](https://user-images.githubusercontent.com/23121882/79051412-68fe6b80-7c30-11ea-826f-9b3b246a3ef7.png)

anntzer commented:
Do you have an implementation, or was that just some inkscaping? :-)

ImportanceOfBeingErnest commented:
No implementation, just 20 seconds of making a screenshot and putting a text box in it.

jklymak commented:
I'd block on a tooltip on the arrow unless it was activated/de-activated with some sort of keyboard shortcut.  We should not be covering data with GUI elements.

anntzer commented:
The tooltip would go away as soon as you move the cursor next to the button.

FWIW here's a scratch implementation for Qt:
```patch
diff --git i/lib/matplotlib/backends/backend_qt5.py w/lib/matplotlib/backends/backend_qt5.py
index 506166e1f..5c6f72966 100644
--- i/lib/matplotlib/backends/backend_qt5.py
+++ w/lib/matplotlib/backends/backend_qt5.py
@@ -711,14 +711,21 @@ class NavigationToolbar2QT(NavigationToolbar2, QtWidgets.QToolBar):
         # The stretch factor is 1 which means any resizing of the toolbar
         # will resize this label instead of the buttons.
         if self.coordinates:
-            self.locLabel = QtWidgets.QLabel("", self)
+            self.locLabel = label = QtWidgets.QLabel("", self)
             self.locLabel.setAlignment(
                 QtCore.Qt.AlignRight | QtCore.Qt.AlignVCenter)
             self.locLabel.setSizePolicy(
                 QtWidgets.QSizePolicy(QtWidgets.QSizePolicy.Expanding,
                                       QtWidgets.QSizePolicy.Ignored))
-            labelAction = self.addWidget(self.locLabel)
-            labelAction.setVisible(True)
+            self.locLabel.setFrameStyle(QtWidgets.QFrame.Raised)
+            super_paintEvent = label.paintEvent
+
+            def paintEvent(event):
+                QtGui.QPainter(label).fillRect(
+                    label.rect(), label.palette().color(label.backgroundRole()))
+                super_paintEvent(event)
+
+            label.paintEvent = paintEvent
 
     def edit_parameters(self):
         axes = self.canvas.figure.get_axes()
@@ -767,6 +774,15 @@ class NavigationToolbar2QT(NavigationToolbar2, QtWidgets.QToolBar):
         self.message.emit(s)
         if self.coordinates:
             self.locLabel.setText(s)
+            self.locLabel.adjustSize()
+            if s:
+                # TODO: Only do this if it would bump into tool buttons.
+                self.locLabel.setFrameShape(QtWidgets.QFrame.StyledPanel)
+            else:
+                self.locLabel.setFrameShape(QtWidgets.QFrame.NoFrame)
+            self.locLabel.move(  # Probably on resize too.
+                self.rect().width() - self.locLabel.rect().width(), 0)
 
     def set_cursor(self, cursor):
         self.canvas.setCursor(cursord[cursor])
```

jklymak commented:
If I'm dragging the cursor over the data to find data values or to do a precise zoom, I don't want data near the cursor covered.

anntzer commented:
Ok, so over the toolbar semms fine?

ImportanceOfBeingErnest commented:
I think so. The widget would only cover up the underlying buttons, if the figure is small, and in all the "usual" cases it would look exactly as if the label was part of the widget layout, right?  I would say that such small aestetical problem is bearable. 
Is it then still possible to verically center-align the thing?

jklymak commented:
Will over the tool bar will cover the axes annotations if they are against the edge of the figure (ie after calling constrained/tight layout)?

ImportanceOfBeingErnest commented:
(I think) we're currently discussing something along the lines of the *right* image in [this comment](https://github.com/matplotlib/matplotlib/pull/17092#issuecomment-612480438).

anntzer commented:
> (I think) we're currently discussing something along the lines of the right image in this comment.

Yes.

> in all the "usual" cases it would look exactly as if the label was part of the widget layout, right? I would say that such small aestetical problem is bearable.

Yes.

> Is it then still possible to verically center-align the thing?

Just a bit more work, but not anything difficult, so sure (it's just measuring toolbar height and manual positioning and yada yada).

jklymak commented:
I guess I prefer that to the left image, but, still kinda looks like a mistake.  I'd be fine if the toolbar were 12 points higher and the output went above...

jklymak commented:
...though I guess it would only cover the toolbar if it were in a small window?  I guess that seems fine.

anntzer commented:
Yes, it's only for tiny windows that this matters.

timhoffm commented:
The sketched tooltip examples look like graphics errors to me. Waiting for the actual tooltip, but I think I don‘t like the covering of the controls. It could be an alternative to collapse all controls into a dropdown toolbutton if there is not enough space.

tacaswell commented:
I'm having a bit of trouble following this discussion.  @ImportanceOfBeingErnest can you comment on if you are happy with where this landed?  If so please merge it for 3.3, if not we can push it to 3.4.

anntzer commented:
rebased

QuLogic commented:
Qt will add a little arrow when the toolbar is overfull, so moving the label there will not cause the same behaviour as in Tk. Whether we should instead move the label into a tooltip or overlay is a different question, but as this brings our backends in line with each, I'm :+1: on this even if we don't answer the broader question yet.

QuLogic commented:
On wx, it uses GTK as its backend, which crops the toolbar as well, so no weird figure resizing either.
